Sqlite tutorial (#6)
realm/unity-examples: tutorials/persistance-comparison-series/Sqlite/sqlite.md

The next time we start the game, we want to load this hit count from the database again. We use the `Start()` function (13) since it only needs to be done when the scene loads. As before, we need to get a hold of the database with an `IDbConnection` (14) and create a new `IDbCommand` (15) to read the data. Since there is only one table and one value, it's quite simple for now. We can just read `all data` by using:  

```sql
"SELECT * FROM HitCountTableSimple"
```

In this case, `SELECT` stands for `read the following values`, followed by a `*` which indicates to read all the data. The keyword `FROM` then specifies the table that should be read from, which is again `HitCountTableSimple`. Finally, we execute this command using `ExecuteReader()` (17) since we expect data back. This data is saved in an [`IDataReader`](https://docs.microsoft.com/en-us/dotnet/api/system.data.idatareader?view=net-6.0), from the documentation:

> Provides a means of reading one or more forward-only streams of result sets obtained by executing a command at a data source, and is implemented by .NET data providers that access relational databases.

> `IDataReader` addresses its content in an index fashion, where the ordering matches one of the columns in the SQL table. So in our case, `id` has index 0, and `hitCount` has index 1. The way this data is read is row by row. Each time we call `dataReader.Read()` (18), we read another row from the table. Since we know there is only one row in the table, we can just assign the `value` of that row to the `hitCount` using its index 1. The `value` is of type `INTEGER` so we need to use [`GetInt32(1)`](https://docs.microsoft.com/en-us/dotnet/api/system.data.idatarecord.getint32?view=net-6.0) to read it and specify the index of the field we want to read as a parameter, `id` being `0` and `value` being `1`.

Reviewer: ```suggestion
`IDataReader` addresses its content in an index fashion, where the ordering matches the one of the columns in the SQL table. So in our case, `id` has index 0, and `hitCount` has index 1. The way this data is read is row by row. Each time we call `dataReader.Read()` (18), we read another row from the table. Since we know there is only one row in the table, we can just assign the `value` of that row to the `hitCount` using its index 1. The `value` is of type `INTEGER` so we need to use [`GetInt32(1)`](https://docs.microsoft.com/en-us/dotnet/api/system.data.idatarecord.getint32?view=net-6.0) to read it and specify the index of the field we want to read as a parameter, `id` being `0` and `value` being `1`.
```

    }

    // Remember to always close the connection at the end.
    dbConnection.Close();
}
```

The first part works basically unchanged by creating a `IDbConnection` (1) and a `IDbCommand` (2) and then reading all rows again with `SELECT *` (3) but this time from `HitCountTableExtended`, finished by actually executing the command with `ExecuteReader()` (4).

For the next part, we now need to read each row (5) and then check which `KeyCode` it belongs to. We grab the `id` from index `0` (6) and the `hits` from index `1` (7) as before. Then, we check the `id` against the `KeyCode` (8) and assign it to the corresponding `hitCount` (9).

Now restart the game and try it out!

## Conclusion

SQLite is one of the options when it comes to persistence. If you've read the previous tutorials, you've noticed that using it might at first seem a bit more complicated than the simple `PlayerPrefs`. You have to learn an additional "language" to be able to communicate with your database. And due to the nature of SQL not being the easiest format to read, it might seem a bit intimidating at first. But the world of databases offers a lot more than can be shown in a short tutorial like this!

> One of the downsides of plain files or `PlayerPrefs` that we've seen was having data in a structured way—especially when it gets more complicated or relationships between objects should be drawn. We looked at JSON as a way to improve that situation but as soon as we need to change the format and migrate our structure, it gets quite complicated. Encryption is another topic that might be important for you—`PlayerPrefs` and `File` are not safe and can easily be read. Those are just some of the areas a database like SQLite might help you achieve the requirements you have for persisting your data.

Reviewer: ```suggestion
One of the downsides of plain files or `PlayerPrefs` that we've seen was having data in a structured way—especially when it gets more complicated or relationships between objects should be drawn. We looked at JSON as a way to improve that situation but as soon as we need to change the format and migrate our structure, it gets quite complicated. Encryption is another topic that might be important for you — `PlayerPrefs` and `File` are not safe and can easily be read. Those are just some of the areas a database like SQLite might help you achieve the requirements you have for persisting your data.
```

# Saving Data in Unity3D Using Sqlite
<b>(Part 4 of the Persistence Comparison Series)</b>

## Introduction

> Our journey of exploring options given to use when it comes persistence in Unity will in this part lead to databases. More specificaclly: Sqlite.

Reviewer 2: Personal preference, but most of the time I've seen it spelled SQLite

Author: Which is actually the correct way. I'll adjust that. 👍

> Sqlite is a C based database that is used in many areas. It has been around for a long time and also found its way into the Unity world. During this tutorial series we have seen options like `PlayerPrefs` in Unity and on the other side `File` and `BinaryWriter`/`BinaryReader` provided by the unterlying .NET framework.

Reviewer: ```suggestion
Sqlite is a C based database that is used in many areas. It has been around for a long time and also found its way into the Unity world. During this tutorial series we have seen options like `PlayerPrefs` in Unity and on the other side `File` and `BinaryWriter`/`BinaryReader` provided by the underlying .NET framework.
```

Here is an overview of the complete series:

- Part 1: [PlayerPrefs](https://github.com/realm/unity-examples/tree/persistence-comparison/tutorials/persistance-comparison-series/PlayerPrefs)
- Part 2: [Files](https://github.com/realm/unity-examples/blob/persistence-comparison/tutorials/persistance-comparison-series/File/file.md)
- Part 3: [BinaryReader and BinaryWriter](https://github.com/realm/unity-examples/blob/persistence-comparison/tutorials/persistance-comparison-series/BinaryWriterReader/binary_reader_writer.md)
- Part 4: SQL *(this tutorial)*
- Part 5: Realm Unity SDK *(coming soon)*
- Part 6: Comparison of all those options

> Similar to the previous parts, this tutorial can also be found in our [Unity examples repository](the https://github.com/realm/unity-examples) on the [persistence-comparison](https://github.com/realm/unity-examples/tree/persistence-comparison) branch.

Reviewer: ```suggestion
Similar to the previous parts, this tutorial can also be found in our [Unity examples repository](https://github.com/realm/unity-examples) on the [persistence-comparison](https://github.com/realm/unity-examples/tree/persistence-comparison) branch.
```

The second part to this is saving changes, which we want to do whenever we register a mouse click. The Unity message for this is `OnMouseDown()` (4). This method gets called every time the `GameObject` that this script is attached to is clicked (with a left mouse click). In this case, we increment the `hitCount` (5) which will eventually be saved by the various options shown in this tutorials series.

## Sqlite

(See `SqliteExampleSimple.cs` in the repository for the finished version.)

Now let's make sure our hit count gets persisted so we can continue playing the next time we start the game.

Sqlite is not included per default in a new Unity project and also not available directly via the Unity package manager. We have to install two components to start using it.

First, head over to [https://sqlite.org/download.html](https://sqlite.org/download.html) and choose the `Precompiled Binaries` for your operating system. Unzip it and add the two files - `sqlite3.def` and `sqlite3.dll` - to the `Plugin` folder in your Unity project.

Then, open a file explorer in your Unity Hub installation directory, head to the following sub directory:

```

> Unity/Hub/Editor/2021.2.11f1/Editor/Data/MonoBleedingEdge/lib/mono/unity

Reviewer 2: A question out of ignorance. This "2021.2.11f1", is it the Unity version?

Author: :D 

It indeed is, yes.

```

In there you will find the file `Mono.Data.Sqlite.dll` which also needs to be moved to the `Plugins` folder in your Unity project. The result when going back to the Editor should look like this:

<img src="images/05_sqlite_in_project_hierarchy.png" alt="Capsule in scene"/>

Now that the preparations are finished we want to add our first script to the capsule. Similar to the `HitCountExample.cs` create a new `C# script` and name it `SqliteExampleSimple`.

When opening it, the first thing we want to do, is importing Sqlite by adding `using Mono.Data.Sqlite;` and `using System.Data;` at the top of the file (1).

> Next we will look at how to save whenever the hit count is changed, which happens during `OnMouseDown()`. Similar to the `FileStream` and the `BinaryWriter` we have seen in the previous parts of this series we need some way to write the data to where it should go. In case of Sqlite that means we need a connection to the database. This is offered by the Sqlite library via the [`IDbConnection`](https://docs.microsoft.com/en-us/dotnet/api/system.data.idbconnection?view=net-6.0) class (2) which represents an open connection to the database. Since we will need a connection for loading the data later on again, we will extract opening a database connection into into another function and call it `private IDbConnection CreateAndOpenDatabase()` (3).

Reviewer: ```suggestion
Next we will look at how to save whenever the hit count is changed, which happens during `OnMouseDown()`. Similar to the `FileStream` and the `BinaryWriter` we have seen in the previous parts of this series we need some way to write the data to where it should go. In case of Sqlite that means we need a connection to the database. This is offered by the Sqlite library via the [`IDbConnection`](https://docs.microsoft.com/en-us/dotnet/api/system.data.idbconnection?view=net-6.0) class (2) which represents an open connection to the database. Since we will need a connection for loading the data later on again, we will extract opening a database connection into another function and call it `private IDbConnection CreateAndOpenDatabase()` (3).
```

Reviewer: > Similar to the `FileStream` and the `BinaryWriter` we have seen in the previous parts of this series we need some way to write the data to where it should go. In case of Sqlite that means we need a connection to the database

I'd just remove this. And go like: In order to save date in Sqlite we first need to open a connection to the database.

Something along these lines.

> In there we first define a name for our database file. I'll just call it `MyDatabase` for now, accordingly the URI should be `"URI=file:MyDatabase.sqlite"` (4). Then we can create a connection this database using `new SqliteConnection(dbUri)` (5) and open it with `dbConnection.Open()` (6).

Reviewer: ```suggestion
In there we first define a name for our database file. I'll just call it `MyDatabase` for now, accordingly the URI should be `"URI=file:MyDatabase.sqlite"` (4). Then we can create a connection to this database using `new SqliteConnection(dbUri)` (5) and open it with `dbConnection.Open()` (6).
```

```cs
using Mono.Data.Sqlite; // 1
using System.Data; // 1
using UnityEngine;

public class SqliteExampleSimple : MonoBehaviour
{
    // Resources:
    // https://www.mono-project.com/docs/database-access/providers/sqlite/

    [SerializeField] private int hitCount = 0;

    void Start() // 13
    {
        // Read all values from the table.
        IDbConnection dbConnection = CreateAndOpenDatabase(); // 14
        IDbCommand dbCommandReadValues = dbConnection.CreateCommand(); // 15
        dbCommandReadValues.CommandText = "SELECT * FROM HitCountTableSimple"; // 16
        IDataReader dataReader = dbCommandReadValues.ExecuteReader(); // 17

>         while (dataReader.Read())

Reviewer: ```suggestion
        while (dataReader.Read()) // 18
```

    {
        hitCount++;

        // Insert hits into the table.
        IDbConnection dbConnection = CreateAndOpenDatabase(); // 2
        IDbCommand dbCommandInsertValue = dbConnection.CreateCommand(); // 9
        dbCommandInsertValue.CommandText = "INSERT OR REPLACE INTO HitCountTableSimple (id, hits) VALUES (0, " + hitCount + ")"; // 10
        dbCommandInsertValue.ExecuteNonQuery(); // 11

        // Remember to always close the connection at the end.
        dbConnection.Close(); // 12
    }

    private IDbConnection CreateAndOpenDatabase() // 3
    {
        // Open a connection to the database.
        string dbUri = "URI=file:MyDatabase.sqlite"; // 4
        IDbConnection dbConnection = new SqliteConnection(dbUri); // 5
        dbConnection.Open(); // 6

        // Create a table for the hit count in the database if it does not exist yet.
        IDbCommand dbCommandCreateTable = dbConnection.CreateCommand(); // 6
        dbCommandCreateTable.CommandText = "CREATE TABLE IF NOT EXISTS HitCountTableSimple (id INTEGER PRIMARY KEY, hits INTEGER )"; // 7
        dbCommandCreateTable.ExecuteReader(); // 8

        return dbConnection;
    }
}
```

> Now we can work with this Sqlite database. Before we can actually add data to the it though, we need to set up a structure. This means, creating and defining tables which is the way most databases are organized. The following screenshot shows the final state we will create in this example.

Reviewer: ```suggestion
Now we can work with this Sqlite database. Before we can actually add data to it though, we need to set up a structure. This means, creating and defining tables which is the way most databases are organized. The following screenshot shows the final state we will create in this example.
```

<img src="images/06_sqlite_tables.png" alt="Sqlite tables"/>

> When accessing or modifying the database we use [`IDbCommand`](https://docs.microsoft.com/en-us/dotnet/api/system.data.idbcommand?view=net-6.0) (6) which represents and SQL statement that can be executed on a database.

Reviewer: ```suggestion
When accessing or modifying the database we use [`IDbCommand`](https://docs.microsoft.com/en-us/dotnet/api/system.data.idbcommand?view=net-6.0) (6) which represents an SQL statement that can be executed on a database.
```

Let's create a new table and define some colums using the following command (7):

```sql
"CREATE TABLE IF NOT EXISTS HitCountTableSimple (id INTEGER PRIMARY KEY, hits INTEGER )"
```

> So, what does this statement mean? First we need to state what we want to do, which is `CREATE TABLE IF NOT EXISTS`, then we neede to name this table, which will just be the same as the script we are working on right now: `HitCountTableSimple`.

Reviewer: ```suggestion
So, what does this statement mean? First we need to state what we want to do, which is `CREATE TABLE IF NOT EXISTS`, then we need to name this table, which will just be the same as the script we are working on right now: `HitCountTableSimple`.
```

> So, what does this statement mean? First we need to state what we want to do, which is `CREATE TABLE IF NOT EXISTS`, then we neede to name this table, which will just be the same as the script we are working on right now: `HitCountTableSimple`.

Reviewer 2: ```suggestion
So, what does this statement mean? First we need to state what we want to do, which is `CREATE TABLE IF NOT EXISTS`, then we need to name this table, which will just be the same as the script we are working on right now: `HitCountTableSimple`.
```

> Last but not least we need to define how this new table is supposed to look like. This is done by naming all columns as a tuple, like `(id INTEGER PRIMARY KEY, hits INTEGER )`, with the column name (the `key`) being the first variable (`id` and `hits`), followed by the type (`INTEGER`) and any additional attributes like `PRIMARY KEY`.

Reviewer: > with the column name (the `key`) being 

I'm confused here by (the `key`). What are you referring to? If database keys, I'm a bit puzzled. Those columns to my knowledge are called `attribute`. What is that you wanted to say?

After assigning this statement as the `CommandText` we need to call `ExecuteReader()` (8) on `dbCommandCreateTable` to run it.

Now back to `OnMouseClicked()`. With the `dbConnection` created we can now go ahead and define another `IDbCommand` (9) to modify the new table we just created and add some data. This time the `CommandText` (10) will be:

```sql
"INSERT OR REPLACE INTO HitCountTableSimple (id, hits) VALUES (0, " + hitCount + ")"
```

> Let's decypher this one too: `INSERT OR REPLACE INTO` adds a new variable to a table or updates it, if it already exists. Next it the table name that we want to insert into, `HitCountTableSimple`. This is followed by a tuple of columns (using their key) that we would like to change, `(id, hits)`. The statement `VALUES (0, " + hitCount + ")` then defines values that should be inserted, also as a tuple. In this case we just choose `0` for the key and use whatever the current `hitCount` is as the value.

Reviewer: ```suggestion
Let's decypher this one too: `INSERT OR REPLACE INTO` adds a new variable to a table or updates it, if it already exists. Next is the table name that we want to insert into, `HitCountTableSimple`. This is followed by a tuple of columns (using their key) that we would like to change, `(id, hits)`. The statement `VALUES (0, " + hitCount + ")` then defines values that should be inserted, also as a tuple. In this case we just choose `0` for the key and use whatever the current `hitCount` is as the value.
```

> Let's decypher this one too: `INSERT OR REPLACE INTO` adds a new variable to a table or updates it, if it already exists. Next it the table name that we want to insert into, `HitCountTableSimple`. This is followed by a tuple of columns (using their key) that we would like to change, `(id, hits)`. The statement `VALUES (0, " + hitCount + ")` then defines values that should be inserted, also as a tuple. In this case we just choose `0` for the key and use whatever the current `hitCount` is as the value.

Reviewer: > ...tuple of columns (using their key)...

> ...for the key and use whatever...

Again, I'm confused by the usage of the word "key".

Opposed to creating the table we execute this command calling `ExecuteNonQuery()` (11) on it.

The difference can be defined as follows:

> ExecuteReader is used for any result set with multiple rows/columns (e.g., SELECT col1, col2 from sometable ). ExecuteNonQuery is typically used for SQL statements without results (e.g., UPDATE, INSERT, etc.).

> ALl that's left to do is properly `Close()` (12) the database.

Reviewer: ```suggestion
All that's left to do is to properly `Close()` (12) the database.
```

After downloading and installing it, all you need to do is `File -> Open Database`, then browse to your Unity project and select the `MyDatabase.sqlite` file. If you then choose the `Table` `HitCountTableSimple`, the result should then look something like this:

<img src="images/07_db_browser.png" alt="Capsule in scene" width="800"/>

Go ahead and run your game, click a couple times on the capsule and check the Inspector for the change. When you then go back to the DB browser and click refresh, the same number should appear in the `value` column of the table.

The next time we start the game, we want to load this hit count from the database again. We use the `Start()` function (13) since it only needs to be done when the scene loads. As before, we need to get a hold of the database with an `IDbConnection` (14) and create a new `IDbCommand` (15) to read the data. Since there is only one table and one value, it is quite simple for now. We can just read `all data` by using:  

```sql
"SELECT * FROM HitCountTableSimple"
```

> In this case, `SELECT` stands for `read the following values`, followed by a `*` which indicates to read all data. The keyword `FROM` than specifie the table that should be read from which is again `HitCountTableSimple`. Finally we execute this command using `ExecuteReader()` (17) since we expect data back. This data is saved in an [`IDataReader`](https://docs.microsoft.com/en-us/dotnet/api/system.data.idatareader?view=net-6.0), from the documentation:

Reviewer: ```suggestion
In this case, `SELECT` stands for `read the following values`, followed by a `*` which indicates to read all the data. The keyword `FROM` then specifies the table that should be read from which is again `HitCountTableSimple`. Finally we execute this command using `ExecuteReader()` (17) since we expect data back. This data is saved in an [`IDataReader`](https://docs.microsoft.com/en-us/dotnet/api/system.data.idatareader?view=net-6.0), from the documentation:
```

> Provides a means of reading one or more forward-only streams of result sets obtained by executing a command at a data source, and is implemented by .NET data providers that access relational databases.

> The way this data is read is row by row. Each time we call `dataReader.Read()` (18) we read another row from the table. Since we know there is only one row in the table we can just assign the `value` of that row to the `hitCount`. The `value` is of type `INTEGER` so we need to use [`GetInt32(1)`](https://docs.microsoft.com/en-us/dotnet/api/system.data.idatarecord.getint32?view=net-6.0) to read it and specify the index of the field we want to read as a parameter, `id` being `0` and `value` being `1`.

Reviewer: Not being that well familiar with `IDataReader`, the appearance of the word "index" threw me off. I had to go to the documentation to understand what it's meant with it.
Is there something that could be done here to briefly brush up on `IDataReader`? I'd be happy with a simple line like: `IDataReader` addresses its content in an index fashion, where the ordering matches that one of the attributes in the SQL table of interest. So in our case `id` has index 0, and `hitCount` has index 1.

Something like this would have made it a lot clearer for me. If I'm the only one, then disregard this.

Finally, let's read those values from the database when restarting the game.

```cs
void Start()
{
    // Read all values from the table.
    IDbConnection dbConnection = CreateAndOpenDatabase(); // 1
    IDbCommand dbCommandReadValues = dbConnection.CreateCommand(); // 2
    dbCommandReadValues.CommandText = "SELECT * FROM HitCountTableExtended"; // 3
    IDataReader dataReader = dbCommandReadValues.ExecuteReader(); // 4

    while (dataReader.Read()) // 5
    {
        // The `id` has index 0, our `value` has the index 1.
        var id = dataReader.GetInt32(0);  // 6
        var hits = dataReader.GetInt32(1); // 7
        if (id == (int)KeyCode.LeftShift) // 8
        {
            hitCountShift = hits; // 9
        }
        else if (id == (int)KeyCode.LeftControl) // 8
        {
            hitCountControl = hits; // 9
        }
        else
        {
            hitCountUnmodified = hits; // 9
        }
    }

    // Remember to always close the connection at the end.
    dbConnection.Close();
}
```

> The first part works basically unchanged by creating a `IDbConnection` (1) and a `IDbCommand` (2) and then read all rows again with `SELECT *` (3) but this time from `HitCountTableExtended`, finished by actually executing the command with `ExecuteReader()`

Reviewer: ```suggestion
The first part works basically unchanged by creating a `IDbConnection` (1) and a `IDbCommand` (2) and then reading all rows again with `SELECT *` (3) but this time from `HitCountTableExtended`, finished by actually executing the command with `ExecuteReader()`
```

(4).

For the next part we now need to read each row (5) and then check which `KeyCode` it belongs to. We grab the `id` from index `0` (6) and the `hits` from index `1` (7) as before. Then we check the `id` against the `KeyCode` (8) and assign it to the corresponding `hitCount` (9).

Now restart the game and try it out!

## Conclusion

> Sqlite is one of the options when it comes to persistence. If you have read the previous tutorials you have noticed that using it might at first seem a bit more complicated than the simple `PlayerPrefs`. You have to learn an additional 'language' to be able to communicate with your database. And due to the nature of SQL no being the easiest format to read, it might seem a bit intimmidating at first.

Reviewer: ```suggestion
Sqlite is one of the options when it comes to persistence. If you have read the previous tutorials you have noticed that using it might at first seem a bit more complicated than the simple `PlayerPrefs`. You have to learn an additional 'language' to be able to communicate with your database. And due to the nature of SQL not being the easiest format to read, it might seem a bit intimidating at first.
```

> But the world of databases offers a lot more than can be shown in a short tutorial like this!

Reviewer: Is it necessary for this to be split in a paragraph of its own?

One of the downsides of plain files or `PlayerPrefs` that we've seen was having data in a structured way. Especially when it gets more complicated or relationships between objects should be drawn. We looked at JSON as a way to improve that situation but as soon as we need to change the format and migrate our structure it gets quite complicated. Encryption is another topic that might be important for you - `PlayerPrefs` and `File` are not safe and can easily be read. Those are just some of the areas a database like Sqlite might help you achieve the requirements you have for persisting your data.

> In the next tutorial we will look at another database, the Realm Unity SDK, which offers similar advantages like Sqlite, while being very easy to use at the same time.

Reviewer: ```suggestion
In the next tutorial we will look at another database, the Realm Unity SDK, which offers similar advantages to Sqlite, while being very easy to use at the same time.
```
